Elements: Merge element style and classname generation to single filter

aaronrobertshaw opened the pull request:
Addresses: https://github.com/WordPress/gutenberg/issues/59462
Related: https://github.com/WordPress/gutenberg/pull/59533

## What?

This is an alternate approach to fixing https://github.com/WordPress/gutenberg/issues/59462 and the further issue noted on the original fix in https://github.com/WordPress/gutenberg/pull/59533.

TL;DR - Prevents other filters from disrupting the element class name generation while also reducing the chance for class name conflicts.

## Why?

While https://github.com/WordPress/gutenberg/pull/59533 fixed the original issue it introduced further albeit reduced issues.

## How?

- Switch elements block support from using `pre_render_block` to `render_block_data`
- Within the new `render_block_data` filter, generate and store both the elements styles and class name
- Update the `render_block` filter to double-check if an elements class name has been added to the block attributes then add it to the block content

## Testing Instructions

Confirm https://github.com/WordPress/gutenberg/issues/59462 is still resolved

1. Edit a post and add a paragraph 
2. Create a link with the paragraph
3. Via the block inspector apply a link color to the paragraph
4. Save and view the frontend confirming the correct link color is shown

Confirm nested applications of the same element styles work correctly:

1. Create some nested blocks containing links
2. Apply a set of element styles to the parent block
3. Apply different styles to a child block
4. Apply the first set of element styles to a grand-child block
5. Confirm the correct display of element styles

<details>
<summary>Example block markup</summary>

```html
<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"var:preset|color|accent-3"}}}},"className":"custom-link-color","layout":{"type":"constrained"}} -->
<div class="wp-block-group custom-link-color has-link-color"><!-- wp:paragraph -->
<p>Welcome to <a href="http://www.wordpress.org">WordPress</a>. This is your first post. Edit or delete it, then start writing!</p>
<!-- /wp:paragraph -->

<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"var:preset|color|accent-4"}}}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group has-link-color"><!-- wp:paragraph -->
<p>Secondary paragraph with <a href="http://test.com">link</a></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"var:preset|color|accent-3"}}}}} -->
<p class="has-link-color"><a href="http://test.test">Third paragraph</a> </p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```
</details>

## Screenshots or screencast <!-- if applicable -->

| Before | After |
|---|---|
| <img width="639" alt="Screenshot 2024-03-04 at 12 09 04 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/31ed8e45-f6a7-4ec8-a6e1-e89aa877cf66"> | <img width="649" alt="Screenshot 2024-03-04 at 12 08 23 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/8d7abcf8-0eeb-429f-b321-78a61f428f3f"> |


In the before screenshot the third paragraph's link color is not applied correctly.

aaronrobertshaw committed:
Elements: Merge element style and classname generation to single filter

Making the class name and style generation occur in the same filter allows the same block data to be used to generate the class regardless of if that data is filtered further elsewhere.

Note: There's still possibility of class name conflict but greatly reduced.

aaronrobertshaw commented:
The test failures are related to the changes in this PR. I'm working on fixes for them now.

aaronrobertshaw committed:
Update the unit tests

ramonjd commented on lib/block-supports/elements.php:
```diff
@@ -80,47 +71,31 @@ function gutenberg_render_elements_support( $block_content, $block ) {
- * @return null
+ * @return array The same parsed block with elements classname added if appropriate.
  */
-function gutenberg_render_elements_support_styles( $pre_render, $block ) {
```
[wp_render_elements_support_styles](https://github.com/WordPress/WordPress/blob/master/wp-includes/block-supports/elements.php#L142) has been a public function since WP 6.0. 

I searched on wpdirectory.net and there doesn't appear to be any usages, but I'm wondering if we should formally deprecate this one to make sure?

aaronrobertshaw replied:
I'm not 100% sure on this one. These filters weren't designed to be reused. They're also not publicly documented as part of our API, and only really useful as a filter for this specific use case.

Given that, I'm leaning towards handling any deprecation requirements in the core backport.

peterwilsoncc replied:
I'd recommend handling it consistently in Gutenberg if the intent is to convert the parameter to `$deprecated` in Core.

It's the kind of thing that can be easily missed when merging to the wp-dev repo so it would be good to avoid any risk.

It's also hooked on to `render_block_data` with `10, 2` so may end up throwing an error. I'm unable to see if the filter `render_block_data` has changed in GB but it's pretty old so is probably only in WP-Dev now days.

andrewserong commented on lib/block-supports/elements.php:
```diff
@@ -200,11 +189,42 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
-add_filter( 'render_block', 'gutenberg_render_elements_support', 10, 2 );
-add_filter( 'pre_render_block', 'gutenberg_render_elements_support_styles', 10, 2 );
+add_filter( 'render_block', 'gutenberg_render_elements_class_name', 10, 2 );
+add_filter( 'render_block_data', 'gutenberg_render_elements_support_styles', 10, 2 );
```
Tiny detail: would it be worth pre-emptively adding `remove_filter()` calls for the WP version of the new function `gutenberg_render_elements_class_name`, e.g.

```
remove_filter( 'render_block', 'wp_render_elements_class_name', 10, 2 );
```

Just so we don't forget to add it in once we get to core backports?

aaronrobertshaw replied:
Good catch, thanks!

aaronrobertshaw committed:
Remove core render_block filter for adding class to markup

tellthemachines commented on lib/block-supports/elements.php:
```diff
@@ -200,11 +189,43 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
+}
+
+/**
+ * Ensure the elements block support class name generated and added to
```
This seems to be a bug; I'd expect it to not be necessary to add the attributes manually to `$block_content`. It might be [this issue](https://core.trac.wordpress.org/ticket/56417) resurfacing. I was meaning to fix it at the time and then forgot all about it 😅

aaronrobertshaw replied:
My inclination was that it was due to the block being static and the markup having been generated at the point it was saved rather than render.

ramonjd commented on lib/block-supports/elements.php:
```diff
@@ -200,11 +189,43 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
+ */
+function gutenberg_render_elements_class_name( $block_content, $block ) {
+	$class_string = $block['attrs']['className'] ?? '';
+	preg_match( '/\bwp-elements-\S+\b/', $class_string, $matches );
```
I'm not sure this will correctly match classname where the target classname is the only or last item in the classlist.


<img width="584" alt="Screenshot 2024-03-04 at 2 20 38 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/4481996a-12a5-4bfd-98c9-d7cc406b85a3">

Maybe a lookbehind?

`/(?<=["\s])(wp-elements-\w+)/`

andrewserong replied:
I thought it was safe in this case because it's only looking at the `className` attribute in `$block['attrs']`, and not the class name within an HTML string?

ramonjd replied:
Ignore me then. 😄

aaronrobertshaw replied:
I might be missing something. The `$block['atts']['className']` value isn't HTML markup. It would only be `test wp-elements-1as32dfdfdf`. The `\b` word bound still looks like it should be ok in the tests I've done.

<img width="522" alt="Screenshot 2024-03-04 at 1 31 09 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/79ff58a5-c3ad-4ee1-af45-c1f7e56430a2">

<img width="323" alt="Screenshot 2024-03-04 at 1 31 12 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/268b1cea-f3eb-455a-8656-bcd0c77610e4">

ramonjd replied:
I resolved this comment already. Thanks for checking. I didn't read it right.

andrewserong reviewed (approved):
This looks like a good follow-up fix to me, thanks for digging in @aaronrobertshaw!

✅ Tested that #59462 is still resolved
✅ Nested link styles are now working correctly

<img width="649" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/c2077699-fbdc-4489-82c6-dba6984c40ce">

There'll likely be a little bit of stuff to deal with in the backport to core process (deprecating the existing function, swapping null coalescing operators for ternaries?), but the approach here looks good to me, and I like the idea of injecting the classname into the attribute, and then ensuring it's output in the render callback/filter.

LGTM! ✨

ramonjd commented:
Manual testing works as described for me too 🚀

peterwilsoncc commented on phpunit/block-supports/elements-test.php:
```diff
@@ -72,7 +72,12 @@ public function test_elements_block_support_class( $color_settings, $elements_st
 		);
 
-		$actual = gutenberg_render_elements_support( $block_markup, $block );
+		// To ensure a consistent elements class name it is generated within a
+		// `render_block_data` filter and stored in the `className` attribute.
+		// As a result the block data needs to be passed through the same
+		// function for this test.
```
```
/*
 * Multiline comments
 * use this format
 */
```

Per WP Coding standards for PHP and JS. https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-inline-comments

A common gotcha is to start wiht `/**` -- please don't as that's used by parsers looking for docblocks

peterwilsoncc commented on lib/block-supports/elements.php:
```diff
@@ -200,11 +189,43 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
+	$tags = new WP_HTML_Tag_Processor( $block_content );
+
+	if ( $tags->next_tag() ) {
+		// Ensure the elements class name set in render_block_data filter is applied in markup.
+		// See `gutenberg_render_elements_support_styles`.
```
As above re comment formatting.

peterwilsoncc reviewed (commented):
Follow up re: Raymon's comment on back-compat. 

And while I was here figured I'd not the coding standards thing.

Sorry for the late notice, as someone who primarily track wp-dev it's easier to track GB commits rather than all the PRs.

andrewserong commented:
I just cherry-picked this PR to the release/17.8 branch to get it included in the next release: 374f65253cf

tellthemachines commented:
> I'd recommend handling it consistently in Gutenberg if the intent is to convert the parameter to $deprecated in Core.

@peterwilsoncc this is just a filter that was added to the `pre_render_block` hook and is no longer needed. Does this type of function need to be deprecated? It's not publicly documented and it's hard to imagine it would be of any use to extenders.

peterwilsoncc commented:
@tellthemachines typically we deprecate even the private functions as historically extenders have used functions & classes whether they're marked public, private or internal. List Table classes prior to [r54378](https://core.trac.wordpress.org/changeset/54378) are the poster child for this but there's other common examples too.

I'm trying to understand how it hooks on to `render_block_data` -- it used to manipulate `$source_block` but is now using `$parsed_block`, is that because the HTML parser Denis and others worked on allows you to avoid re-working the code?

aaronrobertshaw commented:
Thanks for the feedback @peterwilsoncc 👍 

I'll put together two PRs shortly:
1. To clean up the filters here in Gutenberg, including deprecations
2. A backport including this PRs changes, along with the suggested tweaks, deprecations, and code style fixes

> I'm trying to understand how it hooks on to render_block_data

If I follow the question correctly, it isn't clear how the element block supports use to work versus the latest approach in this PR?

Previously, the elements block supports were implemented in two parts:

1. Generating CSS for a block's element styles and storing those in the style engine under a generated class name (`pre_render_block`)
2. Generating that class name again and injecting that into the block's markup  (`render_block`)

I believe this was originally done this way so that the order the CSS was generated resulted in the correct cascade. If the CSS generation happened in the original `render_block` filter only, then the order would have been back to front. With the choice to use `pre_render_block`, that didn't offer the ability to manipulate the block markup to inject the elements CSS class.

To generate a consistent class name between those two filters (`render_block` and `pre_render_block`), the `$block` object was hashed but this broke when the block's data gets filtered within `render_block_data` such as https://github.com/WordPress/gutenberg/pull/59057.

That brings us to this PR and the current approach:

- The work generating the element styles and class name has been moved from the `pre_render_block` filter to `render_block_data`. 
- This allows the class name to be generated once only and stored in the block attributes so it is robust in the face of filtered attributes.
- The previous `pre_render_block` filter didn't provide a means to store the class name or change the markup forcing the switch to `render_block_data` instead.
   - It was an unfortunate oversight of mine to not completely rename the callback functions for cleaner deprecation. My apologies.
- To allow static blocks that already have their markup saved to get the new elements class name, a `render_block` callback is still needed but it now only checks for an elements class within the latest parsed attributes and adds it to the markup if needed.

I hope that helps clarify the situation some 🤞 

If not, I'll get a couple of PRs up addressing the feedback and they might shed further light on the matter.

Thanks again for the feedback and help!

aaronrobertshaw commented:
A very quick and rough draft for updating Gutenberg as discussed above can be found in https://github.com/WordPress/gutenberg/pull/59538. I might not be able to iterate on that PR and a backport until next week however.

getdave commented:
@aaronrobertshaw Thanks for handling the Backport. When done please don't forget the `Backported to WP Core` label 🙏

aaronrobertshaw commented:
Thanks for double-checking @getdave 👍 

I don't believe this is required for 6.5. 

It fixes issues introduced in https://github.com/WordPress/gutenberg/pull/59533 which in turn was intended to fix a bug exposed by https://github.com/WordPress/gutenberg/pull/59057. 

https://github.com/WordPress/gutenberg/pull/59057 only landed in Gutenberg 17.8.

aaronrobertshaw commented:
The changes from this PR and https://github.com/WordPress/gutenberg/pull/59538 are combined in a single backport available in https://github.com/WordPress/wordpress-develop/pull/6214.
